projects/transformers/trainer_mixins/gmp.py

```diff
@@ -0,0 +1,296 @@
+                "gmp/post_prune_bert_sparsity": bert_sparsity,
+            })
+
+        if wandb.run is not None:
+            wandb.log(logs, commit=False)
+            self.control.should_log = True
+
+        self.prune_iteration += 1
+        return train_loss
+
+
+class ThreeStageLRMixin:
+    """
+    This mixin overrides the lr scheduler to use the ThreeStageGMPLR which consists of
+    three stages:
+        1) warmup - linearly ramp up from 0 to max_lr
+        2) pruning - maintain a constant max_lr
+        3) cooldown - decay the learning rate twice (like a StepLR)
+
+    This LR is used for GMP pruning on already pre-trained fully dense network. In
+    contrast, one would most likely not want to use this mixin for GMP style pruning
+    during pre-training. Instead, it may be better to use OneCycle LR to achieve a
+    better eval-loss.
+
+    Params to add to 'trainer_mixin_args':
+    :param warmup_steps: number of warmup steps (i.e. the the stabilization phase)
+    :param cooldown_steps: number of cooldown steps
+    :param max_lr: the lr held constant throughout the pruning phase
+    :param cooldown_gamma: (optional) how much to decay the lr during the cooldown phase
+                           e.g. lr <- lr * cooldown_gamma; used to decay the lr twice;
+                           defaults to 0.1
+
+    Note: This mixin can't be used with the OneCycleLRMixin as they both override
+          `create_scheduler`.
+
+    Note: Notice how this mixin requires the same params for warmup_steps and
+          cooldown_steps as does GradualMagnitudePruningMixin. This is intentional as
+          the phases are meant to align
+    """
+    def __init__(self, *args, **kwargs):
+
+        super().__init__(*args, **kwargs)
+        mixin_args = self.args.trainer_mixin_args
+        assert "warmup_steps" in mixin_args
+        assert "cooldown_steps" in mixin_args
+        assert "max_lr" in mixin_args
+
+        self.max_steps = self.args.max_steps
+        self.warmup_steps = mixin_args["warmup_steps"]
+        self.cooldown_steps = mixin_args["cooldown_steps"]
+        self.max_lr = mixin_args["max_lr"]
+        self.cooldown_gamma = mixin_args.get("cooldown_gamma", 0.1)
+
+        # The LambdaLR (parent class of ThreeStageGMPLR) will multiply this base lr of
+        # 1 times the one at the given step.
+        kwargs["args"].learning_rate = 1
+
+    def create_scheduler(self, num_training_steps: int):
+
+        self.lr_scheduler = ThreeStageGMPLR(
+            self.optimizer,
+            max_lr=self.max_lr,
+            total_steps=self.max_steps,
+            warmup_steps=self.warmup_steps,
+            cooldown_steps=self.cooldown_steps,
+            cooldown_gamma=self.cooldown_gamma,
+        )
+
+
+def get_model(model):
+    """Extract model from DistributedDataParallel."""
+    if isinstance(model, DistributedDataParallel):
+        return model.module
+    else:
+        return model
```

Comment:
The model may be wrapped with deepspeed. It is safer to use HF version of this method:
```
from transformers.modeling_utils import unwrap_model
model = unwrap_model(model)
```
```